### ISSUE TITLE
Add custom model parameters for custom model

### ISSUE THREAD

**Issue by dtrubitsin** [ultralytics/ultralytics#22167]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I create custom model for training with Focal Loss. I want to pass gamma and alpha as parameters via yaml configuration file. When I pass them without pop out – I receive error when cfg check passed. When I pop it out, firstly ultralytics create model with given parameters and then replace them with default as it init model again and read config file without my arguments. NOTE: I use multy-GPU training. 

My question is there any way to init custom model with additional parameters without modifying original library for cfg checks pass?

```
from typing import Any

import torch
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.nn.tasks import DetectionModel
from ultralytics.utils import DEFAULT_CFG, LOGGER, RANK
from ultralytics.utils.loss import FocalLoss, v8DetectionLoss
from ultralytics.utils.tal import make_anchors


class CustomDetectionLoss(v8DetectionLoss):
    """
    Custom loss function for YOLOv8/11 with Focal Loss.
    Inherits from the original v8DetectionLoss and replace BCE with Focal Loss.
    """

    def __init__(self, model, gamma=1.5, alpha=0.25):
        super().__init__(model)
        # Replace BCE with Focal Loss for classification
        self.focal_loss = FocalLoss(gamma=gamma, alpha=alpha)

    def __call__(self, preds: Any, batch: dict[str, torch.Tensor]) -> tuple[torch.Tensor, torch.Tensor]:
        """Calculate the sum of the loss for box, cls and dfl multiplied by batch size. Focal Loss for cls component."""
        loss = torch.zeros(3, device=self.device)  # box, cls, dfl
        feats = preds[1] if isinstance(preds, tuple) else preds
        pred_distri, pred_scores = torch.cat([xi.view(feats[0].shape[0], self.no, -1) for xi in feats], 2).split(
            (self.reg_max * 4, self.nc), 1
        )
        # b, grids, 4
        pred_scores = pred_scores.permute(0, 2, 1).contiguous()
        pred_distri = pred_distri.permute(0, 2, 1).contiguous()

        dtype = pred_scores.dtype
        batch_size = pred_scores.shape[0]
        imgsz = torch.tensor(feats[0].shape[2:], device=self.device, dtype=dtype) * self.stride[0]  # image size (h,w)
        anchor_points, stride_tensor = make_anchors(feats, self.stride, 0.5)

        # targets
        targets = torch.cat((batch["batch_idx"].view(-1, 1), batch["cls"].view(-1, 1), batch["bboxes"]), 1)
        targets = self.preprocess(targets, batch_size, scale_tensor=imgsz[[1, 0, 1, 0]])
        gt_labels, gt_bboxes = targets.split((1, 4), 2)  # cls, xyxy
        mask_gt = gt_bboxes.sum(2, keepdim=True).gt_(0.0)

        # pboxes
        pred_bboxes = self.bbox_decode(anchor_points, pred_distri)  # xyxy, (b, h*w, 4)

        _, target_bboxes, target_scores, fg_mask, _ = self.assigner(
            pred_scores.detach().sigmoid(),
            (pred_bboxes.detach() * stride_tensor).type(gt_bboxes.dtype),
            anchor_points * stride_tensor,
            gt_labels,
            gt_bboxes,
            mask_gt,
        )

        target_scores_sum = max(target_scores.sum(), 1)

        # cls loss - use FOCAL LOSS
        if self.nc > 1:  # cls loss (only if multiple classes)
            # Prepare targets for focal loss
            t = torch.zeros_like(pred_scores, dtype=dtype, device=self.device)
            if fg_mask.sum():
                t[fg_mask] = target_scores[fg_mask].to(dtype)

            # Use focal loss instead BCE
            loss[1] = self.focal_loss(pred_scores, t) / target_scores_sum

        # box loss
        if fg_mask.sum():
            loss[0], loss[2] = self.bbox_loss(
                pred_distri,
                pred_bboxes,
                anchor_points,
                target_bboxes / stride_tensor,
                target_scores,
                target_scores_sum,
                fg_mask,
            )

        loss[0] *= self.hyp.box  # box gain
        loss[1] *= self.hyp.cls  # cls gain
        loss[2] *= self.hyp.dfl  # dfl gain

        return loss.sum() * batch_size, loss.detach()  # loss(box, cls, dfl)


class CustomModel(DetectionModel):
    """Model with FocalLoss integration."""

    def __init__(self, cfg: str = None, ch: int = 3, nc: int = None, verbose: bool = True,
                 gamma: float = 2.0, alpha: float = 0.25):
        super().__init__(cfg, ch, nc, verbose)
        self.gamma = gamma
        self.alpha = alpha

    def init_criterion(self):
        """Initialize the loss criterion for the DetectionModel."""
        return CustomDetectionLoss(self, gamma=self.gamma, alpha=self.alpha)


class CustomTrainer(DetectionTrainer):
    """Trainer to run custom versions of YOLO Detection model."""

    def __init__(self, cfg=DEFAULT_CFG, overrides: dict[str, Any] | None = None, _callbacks=None):
        # Get gamma and alpha from overrides
        self.gamma = overrides.pop("focal_gamma", 2.0) if overrides else 2.0
        self.alpha = overrides.pop("focal_alpha", 0.25) if overrides else 0.25
        LOGGER.info(f"CustomTrainer initialized with Focal Loss: γ={self.gamma}, α={self.alpha}")

        super().__init__(cfg, overrides, _callbacks)

    def get_model(self, cfg: str | None = None, weights: str | None = None, verbose: bool = True):
        """Returns a customized detection model with Focal Loss support."""
        model = CustomModel(
            cfg=cfg or self.model if isinstance(self.model, str) else None,
            ch=self.data["channels"],
            nc=self.data["nc"],
            verbose=verbose and RANK == -1,
            gamma=self.gamma,
            alpha=self.alpha
        )
        if weights:
            model.load(weights)
        return model
```

logs:
```
CustomTrainer initialized with Focal Loss: γ=1.5, α=0.75
WARNING ⚠️ 'save_hybrid' is deprecated and will be removed in the future.
WARNING ⚠️ 'crop_fraction' is deprecated and will be removed in the future.
Ultralytics 8.3.202 🚀 Python-3.11.10 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 4080, 16072MiB)
                                                       CUDA:1 (NVIDIA GeForce RTX 4080, 16072MiB)

CustomTrainer initialized with Focal Loss: γ=2.0, α=0.25
Ultralytics 8.3.202 🚀 Python-3.11.10 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 4080, 16072MiB)
                                                       CUDA:1 (NVIDIA GeForce RTX 4080, 16072MiB)


### Additional

Ultralytics version 8.3.202

**Comment by UltralyticsAssistant** [comment]
👋 Hello @dtrubitsin, thanks for opening this issue and for using Ultralytics 🚀! This is an automated response to help you get support faster. An Ultralytics engineer will follow up shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report related to passing custom parameters through YAML/overrides (e.g., `focal_gamma`, `focal_alpha`) and multi-GPU behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can reproduce the issue. Please include:
-Exact training command used (CLI) or Python snippet calling your `CustomTrainer`
-Your YAML config/overrides showing where `focal_gamma` and `focal_alpha` are defined
-Minimal custom code required to reproduce (trainer/model/loss), ideally in a few lines or a small file
-Full logs from start to the point where parameters reset or the cfg check fails
-Environment details: OS, GPUs, torch/cuda versions, and `ultralytics` version

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Screenshots or short logs are helpful 📎

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify the behavior on the newest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the clear snippet—what you’re hitting is the config alignment rejecting unknown keys, so `focal_gamma`/`focal_alpha` don’t survive into the second DDP init, which is why you see defaults again. Two supported ways to do this without patching the library:

A) Keep custom params out of the validated trainer config: read them, pass a cleaned dict to the base trainer, then persist them on `self.args` so DDP/resume get the same values.

```python
from ultralytics.models.yolo.detect import DetectionTrainer

class CustomTrainer(DetectionTrainer):
    def __init__(self, cfg=None, overrides=None, _callbacks=None):
        ovr = dict(overrides or {})  # don't mutate caller's dict
        self.gamma = float(ovr.pop("focal_gamma", 2.0))
        self.alpha = float(ovr.pop("focal_alpha", 0.25))
        super().__init__(cfg, ovr, _callbacks)
        # persist for DDP/resume and run reproducibility
        self.args["focal_gamma"] = self.gamma
        self.args["focal_alpha"] = self.alpha

    def get_model(self, cfg=None, weights=None, verbose=True):
        from ultralytics.nn.tasks import DetectionModel
        model = CustomModel(cfg, ch=self.data["channels"], nc=self.data["nc"],
                            verbose=verbose, gamma=self.gamma, alpha=self.alpha)
        if weights:
            model.load(weights)
        return model
```

Then call:
```python
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
model.train(trainer=CustomTrainer, data="coco8.yaml", epochs=3, focal_gamma=1.5, focal_alpha=0.75)
```

B) Put your params in the model YAML (not the training YAML) and read them in `CustomModel.__init__`, e.g. `self.yaml.get("focal_loss", {})`; model YAMLs aren’t validated by the trainer config checks.

For reference, see the Advanced Customization guide on subclassing trainers and models, and note that unknown keys are rejected by the trainer config alignment step, which is why filtering before calling the base `__init__` is necessary: check the Advanced Customization guide and the `check_dict_alignment` behavior in the config reference. Also, `FocalLoss` is documented in the loss reference if you need its expected args. 
- [Advanced Customization (engines, custom trainers/models)](https://docs.ultralytics.com/usage/engine/)
- [Trainer config alignment (`check_dict_alignment`)](https://docs.ultralytics.com/reference/cfg/__init__/#check_dict_alignment)
- [FocalLoss reference](https://docs.ultralytics.com/reference/utils/loss/#focalloss)

**Comment by dtrubitsin** [comment]
@glenn-jocher Thank you for help! I go with way "A" and it worked for me with some modifications: 
- different self.args assignment
- remove unknown args before DetectionValidator init

result code:
```python
class CustomModel(DetectionModel):
    """Model with FocalLoss integration."""

    def __init__(self, cfg: str = None, ch: int = 3, nc: int = None, verbose: bool = True,
                 gamma: float = 2.0, alpha: float = 0.25):
        super().__init__(cfg, ch, nc, verbose)
        self.gamma = gamma
        self.alpha = alpha

    def init_criterion(self):
        """Initialize the loss criterion for the DetectionModel."""
        return CustomDetectionLoss(self, gamma=self.gamma, alpha=self.alpha)


class CustomTrainer(DetectionTrainer):
    """Trainer to run custom versions of YOLO Detection model."""

    def __init__(self, cfg=DEFAULT_CFG, overrides: dict[str, Any] | None = None, _callbacks=None):
        # Get gamma and alpha from overrides
        ovr = dict(overrides or {})  # don't mutate caller's dict
        self.gamma = float(ovr.pop("focal_gamma", 2.0))
        self.alpha = float(ovr.pop("focal_alpha", 0.25))

        super().__init__(cfg, ovr, _callbacks)

        # persist for DDP/resume and run reproducibility
        self.args.focal_gamma = self.gamma
        self.args.focal_alpha = self.alpha

        LOGGER.info(f"CustomTrainer initialized with Focal Loss: γ={self.gamma}, α={self.alpha}")

    def get_model(self, cfg: str | None = None, weights: str | None = None, verbose: bool = True):
        """Returns a customized detection model with Focal Loss support."""
        model = CustomModel(
            cfg=cfg or self.model if isinstance(self.model, str) else None,
            ch=self.data["channels"],
            nc=self.data["nc"],
            verbose=verbose and RANK == -1,
            gamma=self.gamma,
            alpha=self.alpha
        )
        if weights:
            model.load(weights)
        return model

    def get_validator(self):
        """Return a DetectionValidator for YOLO model validation."""
        self.loss_names = "box_loss", "cls_loss", "dfl_loss"

        args = copy(self.args)

        # Remove unknown keys from cfg
        del args.focal_alpha
        del args.focal_gamma

        return DetectionValidator(
            self.test_loader, save_dir=self.save_dir, args=args, _callbacks=self.callbacks
        )
```

Full code reference: https://github.com/dtrubitsin/yolo-focal-loss

**Comment by glenn-jocher** [comment]
Looks good—nice refinement; to avoid manually deleting keys, filter `self.args` against the default schema before creating the validator:

```python
from types import SimpleNamespace
from ultralytics.cfg import DEFAULT_CFG_DICT
from ultralytics.models.yolo.detect import DetectionValidator

def _filter_args(ns):
    allowed = set(DEFAULT_CFG_DICT.keys())
    return SimpleNamespace(**{k: getattr(ns, k) for k in allowed if hasattr(ns, k)})

def get_validator(self):
    self.loss_names = ("box_loss", "cls_loss", "dfl_loss")
    return DetectionValidator(self.test_loader, save_dir=self.save_dir,
                              args=_filter_args(self.args), _callbacks=self.callbacks)
```

This keeps custom params for DDP/resume while passing only supported keys to validation; the extension points you used are the intended ones—see the DetectionTrainer get_model/get_validator reference for signatures and behavior in the docs under [DetectionTrainer reference (get_model/get_validator)](https://docs.ultralytics.com/reference/models/yolo/detect/train/#get_validator). If you later prefer to store hyperparameters in the model YAML and read them in `CustomModel.__init__`, the outline is in the [Model YAML Configuration Guide](https://docs.ultralytics.com/guides/model-yaml-config/).
